src/network/utilization.rs

```diff
@@ -42,4 +45,36 @@ impl Utilization {
             }
         }
     }
+    pub fn remove_ip(&mut self, ips: &[Ipv4Addr]) {
+        // might be possible to refactor this part better
+        // i still don't understand the whole borrow/own system very well yet
+        let placeholder = self.connections.clone();
```

Comment:
I can see why you did this. Probably got yelled at by the borrow checker (cannot borrow as mutable because it is also borrowed as immutable)? If so, this is because you tried to remove from the `HashMap` (there's your mutable borrow) while holding a reference to one of its items (there's your immutable borrow).

The API that will allow you to do this without cloning is [`HashMap::entry`](https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.entry).

Reply:
All this being said, if you are doing this filtering in `UIState::update`, you likely won't need any of this. On a high level, this is because you are already copying data (from memory to the screen), so all you need is to tell it what to copy (i.e. a filter). Then there's no mutability issues involved at all.

Personally I think reducing mutability as much as possible is the best practise regardless of language.